File: python/stack.py
from pointer import Pointer


class Stack:
    def __init__(self, mem, stack_buttom):
        self.mem = mem
        self.pos = stack_buttom

    def push(self, data):
        self.pos -= len(data)
        Pointer(self.mem, self.pos, len(data)).write(data)

    def pop(self, size):
        ret = Pointer(self.mem, self.pos, size).read()
        self.pos += size
        return ret

File: stack.rb
require_relative "pointer"

class Stack
    def initialize(mem, rsp)
        @mem = mem
        @rsp = Pointer.new(rsp, 0, 8)
    end
    
    def push(data)
        @rsp.write_int(@rsp.read_int - data.length)
        Pointer.new(@mem, @rsp.read_int, data.length).write(data)
    end
    
    def pop(size)
        ret = Pointer.new(@mem, @rsp.read_int, size).read
        @rsp.write_int(@rsp.read_int + size)
        return ret
    end
end
